Better ImGui raw input and logging

nipkownix opened the pull request:
This PR replaces the old raw input with a solution based on [ReShade's](https://github.com/crosire/reshade) input manager.

It also adds logging using elishacloud's [logging library](https://github.com/elishacloud/Logging).

Logging should be fine, but I'm never sure what to write in log files (which is why we weren't creating log files until now).
@emoose, let me know if you think we should change something about what is being written into it.

About raw input, my main concern is the possibility of it just not working at all in some systems, although I don't think that's going to happen. (gotta make sure to also test in Wine before an eventual merge)

For anyone who wants to help testing, here's a buid:
[re4_tweaks-ImGui_raw-input-v1.zip](https://github.com/nipkownix/re4_tweaks/files/8085479/re4_tweaks-ImGui_raw-input-v1.zip)

Just press F1 and mess with settings. We're mainly looking to make sure mouse + keyboard are working in the configuration menu, and if key combos work. (Hotkey tab)

If this turns out to be as reliable as I hope, we might be able to add an option to use raw input for aiming too.

emoose commented:
Haven't had a deep look yet but looks pretty good on first impression, so this is letting us make use of RawInput now then? I wonder what it's doing differently to what we tried before, guess the way it hooks RegisterRawInputDevices must be involved somehow.

About things to log, first thing that comes to mind is the screen resolution stuff, seems there's still some people having issues with non-60Hz screens (or game window not displaying) on Steam forums, might be helpful if we could log the display modes & which one game chose (maybe only do this if a launch parameter is set though, since display mode list can be pretty big...)

One other thing I've been thinking about is trying to hook the games `OSReport` & `cLog::err` funcs, seems game tries logging quite a bit with those, maybe that'd be useful for letting user know if game files are corrupt at all (not sure how helpful it'd be though, hooking the `XCOMPRESS` func that caused an exception before might be another option too)

nipkownix commented:
> About things to log, first thing that comes to mind is the screen resolution stuff, seems there's still some people having issues with non-60Hz screens (or game window not displaying) on Steam forums, might be helpful if we could log the display modes & which one game chose (maybe only do this if a launch parameter is set though, since display mode list can be pretty big...)

Added as part of the "VerboseLog" option. Looks like this:

```
17964 14:22:38.218 +---------------+---------------+
17964 14:22:38.218 | Display modes:                |
17964 14:22:38.220 | Mode #0 = 640x480 59Hz
17964 14:22:38.220 | Mode #1 = 640x480 60Hz
17964 14:22:38.220 | Mode #2 = 640x480 67Hz
17964 14:22:38.220 | Mode #3 = 640x480 72Hz
17964 14:22:38.220 | Mode #4 = 640x480 75Hz
17964 14:22:38.220 | Mode #5 = 720x400 59Hz
17964 14:22:38.220 | Mode #6 = 720x400 60Hz
17964 14:22:38.220 | Mode #7 = 720x400 59Hz
17964 14:22:38.220 | Mode #8 = 720x400 60Hz
17964 14:22:38.220 | Mode #9 = 720x400 59Hz
17964 14:22:38.220 | Mode #10 = 720x400 60Hz
17964 14:22:38.220 | Mode #11 = 720x400 70Hz
17964 14:22:38.220 | Mode #12 = 720x480 59Hz
17964 14:22:38.221 | Mode #13 = 720x480 60Hz
17964 14:22:38.221 | Mode #14 = 720x576 50Hz
17964 14:22:38.221 | Mode #15 = 720x576 59Hz
17964 14:22:38.221 | Mode #16 = 720x576 60Hz
17964 14:22:38.221 | Mode #17 = 720x576 59Hz
17964 14:22:38.221 | Mode #18 = 720x576 60Hz
17964 14:22:38.221 | Mode #19 = 720x576 59Hz
17964 14:22:38.221 | Mode #20 = 720x576 60Hz
17964 14:22:38.221 | Mode #21 = 800x600 60Hz
17964 14:22:38.221 | Mode #22 = 800x600 72Hz
17964 14:22:38.221 | Mode #23 = 800x600 75Hz
17964 14:22:38.221 | Mode #24 = 832x624 59Hz
17964 14:22:38.221 | Mode #25 = 832x624 60Hz
17964 14:22:38.222 | Mode #26 = 832x624 59Hz
17964 14:22:38.222 | Mode #27 = 832x624 60Hz
17964 14:22:38.222 | Mode #28 = 832x624 59Hz
17964 14:22:38.222 | Mode #29 = 832x624 60Hz
17964 14:22:38.222 | Mode #30 = 832x624 75Hz
17964 14:22:38.222 | Mode #31 = 1024x768 60Hz
17964 14:22:38.222 | Mode #32 = 1024x768 70Hz
17964 14:22:38.222 | Mode #33 = 1024x768 75Hz
17964 14:22:38.222 | Mode #34 = 1152x864 59Hz
17964 14:22:38.222 | Mode #35 = 1152x864 60Hz
17964 14:22:38.222 | Mode #36 = 1152x864 59Hz
17964 14:22:38.222 | Mode #37 = 1152x864 60Hz
17964 14:22:38.222 | Mode #38 = 1152x864 59Hz
17964 14:22:38.223 | Mode #39 = 1152x864 60Hz
17964 14:22:38.223 | Mode #40 = 1152x864 75Hz
17964 14:22:38.223 | Mode #41 = 1152x872 59Hz
17964 14:22:38.223 | Mode #42 = 1152x872 60Hz
17964 14:22:38.223 | Mode #43 = 1152x872 59Hz
17964 14:22:38.223 | Mode #44 = 1152x872 60Hz
17964 14:22:38.223 | Mode #45 = 1152x872 59Hz
17964 14:22:38.223 | Mode #46 = 1152x872 60Hz
17964 14:22:38.223 | Mode #47 = 1152x872 75Hz
17964 14:22:38.223 | Mode #48 = 1280x720 50Hz
17964 14:22:38.223 | Mode #49 = 1280x720 59Hz
17964 14:22:38.223 | Mode #50 = 1280x720 60Hz
17964 14:22:38.223 | Mode #51 = 1280x800 60Hz
17964 14:22:38.223 | Mode #52 = 1280x1024 60Hz
17964 14:22:38.224 | Mode #53 = 1280x1024 75Hz
17964 14:22:38.224 | Mode #54 = 1366x768 60Hz
17964 14:22:38.224 | Mode #55 = 1440x900 60Hz
17964 14:22:38.224 | Mode #56 = 1600x900 60Hz
17964 14:22:38.224 | Mode #57 = 1680x1050 60Hz
17964 14:22:38.224 | Mode #58 = 1920x1080 23Hz
17964 14:22:38.224 | Mode #59 = 1920x1080 24Hz
17964 14:22:38.224 | Mode #60 = 1920x1080 25Hz
17964 14:22:38.224 | Mode #61 = 1920x1080 25Hz
17964 14:22:38.224 | Mode #62 = 1920x1080 29Hz
17964 14:22:38.224 | Mode #63 = 1920x1080 30Hz
17964 14:22:38.224 | Mode #64 = 1920x1080 29Hz
17964 14:22:38.224 | Mode #65 = 1920x1080 30Hz
17964 14:22:38.225 | Mode #66 = 1920x1080 50Hz
17964 14:22:38.225 | Mode #67 = 1920x1080 59Hz
17964 14:22:38.225 | Mode #68 = 1920x1080 60Hz
17964 14:22:38.225 +---------------+---------------+
```

Something weird I've noticed is that, on 1920x1080 modes, every refresh rate lower than 60 is coming from my secondary monitor, not my primary one. Shouldn't we only be querying the primary monitor?

>E: Maybe logging info about users language/keyboard locale/game language settings could help us look into the EnglishKeyFallback stuff too, not sure if we can access game language on startup though, might need to wait till after game inits or something.

Added language/keyboard locale info as part of ``VerboseLog`` too. Looks like this:

```
17964 14:22:37.962 +------------------+------------------+
17964 14:22:37.962 | Keyboard/locale info:               |
17964 14:22:37.962 | KeyboardLayout: "en-US"
17964 14:22:37.962 | UserDefaultLCID: "en-US"
17964 14:22:37.962 | UserDefaultLCID: "en-US"
17964 14:22:37.962 | UserDefaultUILanguage: "en-US"
17964 14:22:37.962 | SystemDefaultUILanguage: "en-US"
17964 14:22:37.962 | UserDefaultLangID: "en-US"
17964 14:22:37.962 | SystemDefaultLangID: "en-US"
17964 14:22:37.962 +------------------+------------------+
```

Game language is missing though. I'll see if I can add it later.

>One other thing I've been thinking about is trying to hook the games OSReport & cLog::err funcs, seems game tries logging quite a bit with those, maybe that'd be useful for letting user know if there's any corruption in their game data at all (not sure how helpful it'd be though, hooking the XCOMPRESS func that caused an exception before might be a better option for this)

The trainer I mentioned the other day actually does this, and passes the logging to ``cout``. It is interesting, but I don't think it is too useful for us? Might be wrong though.

nipkownix commented:
>  if we could log the display modes & **_which one game chose_**

Forgot about this part. Added now:

`33724 11:22:51.273 Changing game resolution to: 1280x720 60Hz`

Also added the log about which game language is being used:

`33812 11:29:05.906 Game language: Traditional Chinese`

nipkownix commented:
@5963teranigma @ElTioRata @kailip7 @Zaphero1

I'm tagging all of you since y'all mentioned mouse acceleration being an issue.
Here's a test implementation of raw mouse input for mouse turning and aiming:

[dinput8-raw_mouse_aiming-v1.zip](https://github.com/nipkownix/re4_tweaks/files/8102526/dinput8-raw_mouse_aiming-v1.zip)

Very basic, quickly thrown together just so we can determine whether or not adding this feature will help us with mouse acceleration issues at all.
It is enabled by default.
Please let me know the results. Thanks!

emoose commented:
I posted a link to that build in a Steam thread about mouse issues: https://steamcommunity.com/app/254700/discussions/0/3187989286457651007/

>Just tried it out. I do notice a difference with Modern aiming, though the lack of horizontal aim sway (the original issue) is still present. Classic aiming doesn't seem to be affected, and still has its interesting reverse mouse acceleration issue.

emoose commented:
Hmm, wonder if it might be useful to log all the INI settings to log file too? I could imagine some people would post log without thinking of uploading INI too, might help us out if eg. someone crashing was using `EnableDebugMenu` (which is known to cause crashes, or at least was, haven't noticed any recently...)

Also if it isn't already, some way to log the git commit hash would probably be good too.

nipkownix commented:
Good thinking. I'll add those soon.

nipkownix commented:
Added dll version logging now, looks like this:

``31732 00:51:09.281 Starting re4_tweaks v1.7.4.0-da0ee65e-raw-input``

Couldn't find a clean way to log settings though, other than manually logging each ``cfg.*`` var. Hmm.

nipkownix commented:
Settings logging:

````
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 | Setting                        | Value           |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + DISPLAY------------------------+-----------------+
32468 11:27:19.455 | FOVAdditional                  |        0.000000 |
32468 11:27:19.455 | FixUltraWideAspectRatio        |            true |
32468 11:27:19.455 | DisableVsync                   |           false |
32468 11:27:19.455 | FixDisplayMode                 |            true |
32468 11:27:19.455 | CustomRefreshRate              |              -1 |
32468 11:27:19.455 | RestorePickupTransparency      |            true |
32468 11:27:19.455 | DisableBrokenFilter03          |            true |
32468 11:27:19.455 | FixBlurryImage                 |            true |
32468 11:27:19.455 | DisableFilmGrain               |            true |
32468 11:27:19.455 | EnableGCBlur                   |            true |
32468 11:27:19.455 | EnableGCScopeBlur              |            true |
32468 11:27:19.455 | WindowBorderless               |           false |
32468 11:27:19.455 | WindowPositionX                |              -1 |
32468 11:27:19.455 | WindowPositionY                |              -1 |
32468 11:27:19.455 | RememberWindowPos              |           false |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + MISC---------------------------+-----------------+
32468 11:27:19.455 | WrappedDllPath                 |                 |
32468 11:27:19.455 | AshleyJPCameraAngles           |           false |
32468 11:27:19.455 | AllowSellingHandgunSilencer    |            true |
32468 11:27:19.455 | AllowMafiaLeonCutscenes        |            true |
32468 11:27:19.455 | SilenceArmoredAshley           |           false |
32468 11:27:19.455 | SkipIntroLogos                 |           false |
32468 11:27:19.455 | EnableDebugMenu                |           false |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + CONTROLLER---------------------+-----------------+
32468 11:27:19.455 | ControllerSensitivity          |        1.000000 |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + MOUSE--------------------------+-----------------+
32468 11:27:19.455 | UseMouseTurning                |            true |
32468 11:27:19.455 | TurnSensitivity                |        1.000000 |
32468 11:27:19.455 | FixSniperZoom                  |            true |
32468 11:27:19.455 | FixSniperFocus                 |            true |
32468 11:27:19.455 | FixRetryLoadMouseSelector      |            true |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + FRAME RATE---------------------+-----------------+
32468 11:27:19.455 | FixFallingItemsSpeed           |            true |
32468 11:27:19.455 | FixQTE                         |            true |
32468 11:27:19.455 | FixAshleyBustPhysics           |            true |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + MEMORY-------------------------+-----------------+
32468 11:27:19.455 | AllowHighResolutionSFD         |            true |
32468 11:27:19.455 | RaiseVertexAlloc               |            true |
32468 11:27:19.455 | RaiseInventoryAlloc            |            true |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + HOTKEYS------------------------+-----------------+
32468 11:27:19.455 | ConfigMenu                     |              F1 |
32468 11:27:19.455 | Console                        |              F2 |
32468 11:27:19.455 | DebugMenu                      |         CTRL+F3 |
32468 11:27:19.455 | MouseTurningModifier           |             ALT |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + WARNING------------------------+-----------------+
32468 11:27:19.455 | IgnoreFPSWarning               |           false |
32468 11:27:19.455 +--------------------------------+-----------------+
32468 11:27:19.455 + IMGUI--------------------------+-----------------+
32468 11:27:19.455 | FontSize                       |        1.000000 |
32468 11:27:19.455 | DisableMenuTip                 |           false |
32468 11:27:19.455 +--------------------------------+-----------------+
````

Logging seems pretty solid now. Hopefully raw input is too?

kailip7 commented:
> @5963teranigma @ElTioRata @kailip7 @Zaphero1
> 
> I'm tagging all of you since y'all mentioned mouse acceleration being an issue. Here's a test implementation of raw mouse input for mouse turning and aiming:
> 
> [dinput8-raw_mouse_aiming-v1.zip](https://github.com/nipkownix/re4_tweaks/files/8102526/dinput8-raw_mouse_aiming-v1.zip)
> 
> Very basic, quickly thrown together just so we can determine whether or not adding this feature will help us with mouse acceleration issues at all. It is enabled by default. Please let me know the results. Thanks!

I just tested this with Modern settings, 800dpi and 1 ingame sens and it feels like there's no acceleration at all, except when I swipe the mouse quickly back to the starting point, sometimes (it's rather inconsistent) there appears to be a minor degree of negative acceleration. I don't know why that is. Either way, the mouse feel improved considerably and is a lot more consistent and positive acceleration is gone.

The one thing to note is that I think with this fix, ingame sens must have some sort of adjustment, since I'm using 800dpi and the lowest possible sens (1) does not even use a full swipe of my mousepad for a 360. Considering 400-1600dpi range is the usual range most people use, I think ingame sensitivities as they are with the fix are too high for practical usage and proper granularity.

I never noticed it but now that emoose brought that up with that steam post, it is true that Modern settings have no horizontal sway. If that's fixable by adding horizontal sway back, I wonder: would it be possible to hypothetically remove it, as well, and make it an on/off option? I'm aware doing so could kind of make the game easier but I feel like there's no harm in offering the choice, right? Especially when playing with a controller, I always felt that the sway was unnecessary as controllers are already pretty bad for pointing at things, so that always annoyed me back in the day.

kailip7 commented:
So, this is weird... The camera mouse movement seems to be perfectly consistent, but the aiming mouse movement, although improved, still seems a bit off to me, and I'm sure there's negative acceleration now. Trying to think of a way to properly and consistently test and report this.

nipkownix commented:
>  I'm sure there's negative acceleration now.

It is possible that the positive accel from the original implementation was "balancing out" some negative accel that could be present in the actual aiming funcs.

My MouseTurning feature doesn't go through any other game functions, so there shouldn't be any type of accel there, at least.

kailip7 commented:
https://www.youtube.com/watch?v=cIhRgFAWoDo

There, from my testing you can see Leon's head always goes back to roughly the same spot without negative accel when not aiming, but when aiming there is consistent negative acceleration, which is weird.

> It is possible that the positive accel from the game's original implementation was "balancing out" some negative accel that could be present in the actual aiming funcs.
> 
> My MouseTurning feature doesn't go through any other game functions, so there shouldn't be any type of accel there, at least.

That makes a lot of sense, I suppose.

> Ah yeah, the in-game sens config wasn't affecting the sensitivity at all in that build I posted. As I said, that was just a quick test. But since it appears to help, I'm implementing a polished up version with sensitivity support and all, for further testing.

That also makes a lot of sense xD I didn't even notice that in-game sens wasn't working, I just lowered it down and assumed that it was too high.

nipkownix commented:
Ohhh, I've just noticed something very interesting. Seems the game does some crazy speed calculations when aiming, probably to make movement on joysticks feel more "natural" or whatever, but when they made the Wii port, they actually added an ``if`` statement (checking the same ``WPadMode`` @emoose mentioned a few weeks ago) that disables _all_ of this nonsense. Forcing that function to think we are using a Wii-mote makes mouse aiming _very_ precise and responsive. I'm actually impressed with how nicely that + raw input seems to work together. Amazing! Will provide a test build soon.

kailip7 commented:
> Ohhh, I've just noticed something very interesting. Seems the game does some crazy speed calculations when aiming, probably to make movement on joysticks feel more "natural" or whatever, but when they made the Wii port, they actually added an `if` statement (checking the same `WPadMode` @emoose mentioned a few weeks ago) that disables _all_ of this nonsense. Forcing that function to think we are using a Wii-mote makes mouse aiming _very_ precise and responsive. I'm actually impressed with how nicely that + raw input seems to work together. Amazing! Will provide a test build soon.
> 
> Edit: hmm, it seems to also get rid of the aiming "sway" people mentioned before. Not sure how to keep that in.

That's great! Even if there's no way to keep the sway, I think that's a very desirable toggle option for re4_tweaks to have. I would love to play the game with that :P would you mind sending that build my way so that I can try it out?

Also, let me know if you need any help with testing. I'll be glad to help.

nipkownix commented:
> would you mind sending that build my way so that I can try it out?

Sure, I was just cleaning things up.
[dinput8-raw_mouse_aiming-v2.zip](https://github.com/nipkownix/re4_tweaks/files/8104917/dinput8-raw_mouse_aiming-v2.zip)

You can now turn it on/off using the F1 menu during gameplay to compare the old and new behaviors.
Let me know how it goes.

kailip7 commented:
Testing with this newest build, there's no sway like you said but I experienced the same amount of negative acceleration when aiming, doing the same test I did in the video I sent earlier today. So there's still something causing a negative acceleration effect. Weird.

nipkownix commented:
Hmm, doesn't happen here anymore.
Before it happened every once in a while, but nothing like in your video either.
I wonder what else could be different here.

kailip7 commented:
Wait a second, I think this may be FPS related.

I was testing on 30fps and tried on 60fps now, and it seems there's no negative acceleration then. Huh.

nipkownix commented:
Oh? Strange.
Can you check if the first build I posted + 60 fps has mouse accel for you or not?

kailip7 commented:
Just tried that, and for v1 it has negative accel on both 30fps and 60fps modes. v2 only has accel on 30fps mode.

nipkownix commented:
Interesting. So there's still something else going on. Might be ``pG->deltaTime`` related or something.

kailip7 commented:
Yes, sensitivity is working correctly. I had to turn it up immediately since it was turned to 1 previously, it's working fine now.

kailip7 commented:
Ok, so I had a suspicion and tried something out.

I turned up the mouse DPI to see if I could trigger the negative accel because I suspected it was related to cursor speed somehow, and I can't move my mouse very fast in very consistent ways.

So I bumped my DPI from 800 to 1400 and I could trigger the negative accel consistently on 60fps as well.

But why could I trigger it on 800dpi 60fps in v1 but not in v2? I have no idea. It's very strange.

nipkownix commented:
I see. Thanks for the info. I'll increase my DPI later and see if I can reproduce the issue.

nipkownix commented:
Hmm, just tried with 3000 DPI and couldn't trigger the issue with v2 on 60 fps, sadly.
In other news, I thinkg I found precisely what in ``PlWepLockCtrl`` was causing the acceleration issues (in there, at least). So no need to get rid of the sway, I guess.

nipkownix commented:
Build with the change I mentioned in the last post:
[dinput8-raw_mouse_aiming-v3.zip](https://github.com/nipkownix/re4_tweaks/files/8105297/dinput8-raw_mouse_aiming-v3.zip)

Maybe that ends up fixing it for you, who knows :p

Tested with 4800 DPI and didn't notice neither negative or positive acceleration.

nipkownix commented:
So, turns out the horizontal sway is missing because the game locks the camera to the aiming position when the modern mouse setting is enabled.

I've added an option to remove this lock. This means people can now use the modern setting while also having the only benefit that was available in the Classic setting, which is keeping the original camera behavior (+ keeping the sway).

Classic mode is now fully obsolete :p

emoose commented:
Awesome work! Might be nice if we can find a way to change the default to Modern as well, since game seems to choose Classic when config.ini doesn't exist (and I've seen a lot of streamers just start the game without going through options at all >.>)

nipkownix commented:
Well, current behavior is enabling raw input by default, and that forces the aiming mode to Modern. Seemed like the best approach to me as there's really no point in using the new raw mouse work without removing the massive reverse acceleration as well.

kailip7 commented:
> Build with the change I mentioned in the last post: [dinput8-raw_mouse_aiming-v3.zip](https://github.com/nipkownix/re4_tweaks/files/8105297/dinput8-raw_mouse_aiming-v3.zip)
> 
> Maybe that ends up fixing it for you, who knows :p
> 
> Tested with 4800 DPI and didn't notice neither negative or positive acceleration.

I tested with 800DPI and 1400DPI, 30fps and 60fps, couldn't get the same negative acceleration behavior to happen even once.

I think whatever you did fixed it. :)

> So, turns out the horizontal sway is missing because the game locks the camera to the aiming position when the modern mouse setting is enabled.
> 
> I've added an option to remove this lock. This means people can now use the modern setting while also having the only benefit that was available in the Classic setting, which is keeping the original camera behavior (+ keeping the sway).
> 
> Classic mode is now fully obsolete :p

Do you think in addition to this fix you could also add an option to completely disable the sway?

Anonanon commented:
I really want to use [dinput8-raw_mouse_aiming-v3.zip](https://github.com/nipkownix/re4_tweaks/files/8105297/dinput8-raw_mouse_aiming-v3.zip) but none of these releases let me save my dinput8.ini settings, and I have checked the config is not read-only. I also noticed while using this new version, the game will no longer recognize my global Windows hotkeys (I have one to automatically center the window), an accidental benefit when re4tweaks UI is opened. Both problems go away when I revert to the main release 1.7.4.

emoose commented:
@Anonanon hm that's strange, not sure why it wouldn't write to INI, nothing has changed around that lately AFAIK.
Not sure why hotkeys would get disabled neither, some input stuff has changed lately though...

You could probably change WindowPositionX/WindowPositionY to center the window instead, but that'd require knowing the exact X/Y you want to use (and changing resolution would also uncenter it then too... maybe a WindowCentered option would be nice in future)

nipkownix commented:
> but none of these releases let me save my dinput8.ini settings

Hmm, can you upload your dinput8.log so we can take a look?

> the game will no longer recognize my global Windows hotkeys (I have one to automatically center the window)

Global Windows hotkeys? Where do you set those up? Never heard of it. Unless you mean you use some program like AutoHotkey for those, in which case I'd ask you to upload the AHK script so I can try it here as well.

kailip7 commented:
> > but none of these releases let me save my dinput8.ini settings
> 
> Hmm, can you upload your dinput8.log so we can take a look?

I tried it myself and I can't get re4_tweaks to load the settings properly with that build. dinput8.ini has the settings I saved, but the game launches with defaults on and dinput8.log seems to show that:

[dinput8.log](https://github.com/nipkownix/re4_tweaks/files/8111774/dinput8.log)

nipkownix commented:
Oof, you're right. With a little investigation I've found that this issue happens if I ``#include <Logging/Logging.h>`` inside ``Settings.cpp`` for the settings logging. What could even be incompatible between these the logging and ini libraries, though? ._.

kailip7 commented:
> Oof, you're right. With a little investigation I've found that this issue happens if I `#include <Logging/Logging.h>` inside `Settings.cpp` for the settings logging. What could even be incompatible between the logging and ini libraries, though? ._.
> 
> Edit: moved the settings log to dllmain. Kinda ugly, but.. meh.
> 
> Build here: [dinput8-raw_mouse_aiming-v4.zip](https://github.com/nipkownix/re4_tweaks/files/8111857/dinput8-raw_mouse_aiming-v4.zip)

This seems to have fixed it for the most part, but for some reason the new mouse options aren't being properly loaded.

First I checked only UseRawMouseInput and left UnlockCameraFromAim unchecked, then saved. Upon restart, both were unchecked. The ini file had the values I saved, though:

> UseRawMouseInput= True
> UnlockCameraFromAim= False

Then I checked both and saved, and upon restart UseRawMouseInput was checked but not UnlockCameraFromAim. Again, the .ini showed the values I saved:

> UseRawMouseInput= True
> UnlockCameraFromAim= True

Everything else that I tested saved properly, I tried with disabling vsync and skipping intro logos, they both saved just fine so I'm assuming that the problem is now with the new options only.

nipkownix commented:
Pff, weird. The code for loading those is there. I wonder if the logging library is still causing some weirdness with the ini.
Will take a look later, I guess. Pretty annoying though.

kailip7 commented:
> ~Pff, weird. The code for loading those is there. I wonder if the logging library is still causing some weirdness with the ini. Will take a look later, I guess. Pretty annoying though.~
> 
> Edit: wait, looks like I'm loading a setting into the wrong var. Oops.
> 
> This one should be working fine: [dinput8-raw_mouse_aiming-v5.zip](https://github.com/nipkownix/re4_tweaks/files/8112009/dinput8-raw_mouse_aiming-v5.zip)

It's loading the saved settings correctly now, that fixed it. 😄 

Until the next stable release is out I'll be playing through the game with this build. Again, amazing work with those fixes, I really appreciate it as bad mouse input is something that drives me insane with problematic PC ports 🥲

Anonanon commented:
I was going to reply to you with my log but it seems [dinput8-raw_mouse_aiming-v5.zip](https://github.com/nipkownix/re4_tweaks/files/8112009/dinput8-raw_mouse_aiming-v5.zip) fixed settings not loading as well as Raw Mouse Input not turning on. As for Windows global hotkeys not working anymore compared to 1.7.4, it doesn't matter now because I found another workaround which is to bring up Steam Overlay and then use my hotkey there. I didn't want to use the WindowPositionX/WindowPositionY option because I believe the game is missing one or possibly couple columns of pixel, so as the village background scrolls along it looks like there is a seam or screen tear. Maybe this is an issue for another topic?

nipkownix commented:
I noticed that issue you mentioned with the seam as well. @emoose any idea what could be causing this?

emoose commented:
Hm I only get seams when DisableVsync is set, not sure if there's any way to remove them while using that, could be something to do with the way it limits to 60FPS, maybe needs to be more accurate or something.

nipkownix commented:
I see. The seams we're talking about aren't screen tearing, though. Its almost as if the game window is smaller than the render resolution and an entire column of pixels is removed or something.

You actually saw it yourself once, and commented about it in the LAA PR (if I'm not mistaken)

emoose commented:
My memory is terrible heh, vaguely remember something like that but can't find much about it now - also tried checking with windowed/fullscreen but don't seem to get any seams now besides the tearing I mentioned..
I did update to Win11 recently though, maybe the curved borders are hiding it or something.

If you remove re4_tweaks DLL does that get rid of seam for you? My first thought was maybe our vertex buffer change is doing something, since that seemed to affect positions of stuff on screen before IIRC.

nipkownix commented:
I also thought about the vertex buffer being the cause, but I remember ruling that out back then. I remember it being related to the borderless option as well, so might be worth it trying with that on?

emoose commented:
Tried with borderless too, no difference for me, zooming around screenshot shows it's all being fit in there fine:
![hm](https://user-images.githubusercontent.com/1755499/155111410-8d805e97-37cc-4cea-841a-8901847df6ba.png)

Not really sure what else to try, had a go with both 1.0.6 & 1.1.0 and both seem to work fine for me now (I did have some scaling issue with 1.0.6 before IIRC, but seems to be gone now).
Could you post a pic of what it shows for you?

nipkownix commented:
Hm, turns out I cant reproduce it anymore either. I've recently upgraded to Win 11 too, so maybe that changed something for us, idk.

I'm guessing I can go ahead and merge this PR then. Seems raw Input device registration is working fine, which was my biggest fear of breaking.

Danixxxxx commented:
Man, you guys work quick. Just tried the latest test build and the aim already feels SO much smoother, amazing! I can finally make those slow, precise movements in third-person that make a difference on Professional. So not only did 'Modern' have positive acceleration, it still had some negative accel as well? Jesus Capcom. Can't wait for the proper release!

btw maybe it's just me, but it feels to me that aiming up and down isn't as quick as aiming left and right in third person. I compared it to scoped rifle aiming and the sens balance doesn't seem the same. Do you know if the vertical/horizontal sens for third-person is 1:1 or if you can change them separately in any way? Either way, I would be greatly satisfied with this tweak as is. Thanks for your legendary work guys 👍

nipkownix commented:
> btw maybe it's just me, but it feels to me that aiming up and down isn't as quick as aiming left and right in third person. I compared it to scoped rifle aiming and the sens balance doesn't seem the same. Do you know if the vertical/horizontal sens for third-person is 1:1 or if you can change them separately in any way?

The game appears to apply some different sensitivity to the Y aiming, making it slower, so I tried to compensate for that when calculating the raw input delta.

Y:
``*(int32_t*)(ptrMouseDeltaY) = -(int32_t)((_input->raw_mouse_delta_y() / 6.0f) * g_MOUSE_SENS());``

X:
``*(int32_t*)(ptrMouseDeltaX) = (int32_t)((_input->raw_mouse_delta_x() / 10.0f) * g_MOUSE_SENS());``

I think it feels pretty solid right now, though. Maybe I can figure out a way to improve this in the future.

@emoose, that weird line issue we were talking about here came back for me, but a little different this time:

https://user-images.githubusercontent.com/33870163/155891909-ec54c356-da4a-4e89-a195-130befa583ea.mp4

Seems now there's multiple "lines" of missing pixels or something. You can spot them easily with the moving background.

Settings (maybe related?)
```
10516 16:51:42.301 | WindowBorderless               |           false |
10516 16:51:42.301 | WindowPositionX                |              18 |
10516 16:51:42.301 | WindowPositionY                |              45 |
10516 16:51:42.301 | RememberWindowPos              |            true |
```

Not sure if it is worth opening an issue to track this. Hmm.

nipkownix commented:
@emoose, found out the issue. The window size is actually smaller than the game's resolution when using those patches:

![image](https://user-images.githubusercontent.com/33870163/156928199-6c89d084-663d-40ad-bbac-8d2fe8e92b08.png)

Not sure why, I haven't really looked into it yet.

emoose commented:
Is that only when using WindowPositionX/Y? Hm, if they're using a RECT to store position then maybe we're only updating top/left without also adding to bottom/right as well, I see something like that inside GXInit but not sure how it's used yet.

nipkownix commented:
Seems to be caused by this:

```C
		// nop MoveWindow
		pattern = hook::pattern("53 51 52 53 53 50");
		injector::MakeNOP(pattern.get_first(6), 6);
```

Apparently MoveWindow does more than just.. moving the window ._.
